### PR TITLE
Volume: Add auto provision flag to brick info

### PR DESCRIPTION
**Pull request by rafikc30**
Brick should be able to identify whether it is auto provisioned
or not. This is required for cases like

* Mounting of a brick in glusterd restart
* Umounting of a brick on volume stop/delete
* delete brick directoroy when volume is deleted
* delete vg/lv during volume delete
* To smoothly handle replace brick/reset brick on auto provisioned bricks

Signed-off-by: Mohammed Rafi KC <rkavunga@redhat.com>

### CODE REVIEW

**Comment by rafikc30** [comment]
With the above approach,  we can save changes in req struct. But IMO, It cannot be in metadata field for permanent. Given that auto-provision is a brick specific feature, it is not a metadata information for a volume.

Assume a case where we have two bricks. Initially, both are self-provisioned. Then with a replace brick operation lets say we changed one brick to a normal brick. Then the volume metadata field won't make sense, as we have one brick as self-provisioned and one not. Going little further if we do replace brick on the brick as well, then we have to remove the metadata field.

**Comment by phlogistonjohn** [comment]
I'm going to complain about the name 'SelfProvision' because I find "self" a bit ambiguous to the reader. People already talk about things as auto provisioned, as in the title of your pr, why not 'AutoProvisioned' ? 

I'm also frequently worried about upgradability and the nature of fields perhaps not always making sense as a boolean forever. In that vein, one could consider a new type like:
```golang
type ProvisionState string
const (
  ManuallyProvisioned ProvisionState = ""
  AutoProvisioned ProvisionState = "auto"
  MagicProvisioned ProvisionState = "magic-future-thing"
)
```
But I'm willing to be told this is a case of YAGNI.


> Assume a case where we have two bricks. Initially, both are self-provisioned. Then with a replace brick operation lets say we changed one brick to a normal brick. Then the volume metadata field won't make sense, as we have one brick as self-provisioned and one not. Going little further if we do replace brick on the brick as well, then we have to remove the metadata field.

I think this operation should simply not be allowed. Either the brick is replaced through device management functions or a special case of replace that understands the auto-provisioning mechanisms. There simply shouldn't be a way to mix auto-provisioned and manually-provisioned bricks on one volume IMO.

**Comment by rafikc30** [comment]
@phlogistonjohn I agree that AutoProvisioned name suits better. And also instead of having a boolean field, we could have a more meaningful struct that can scale better in future.

IMO, Provisioning a brick is a brick related feature, not a volume field. We can do add/replace/reset brick through a different channel for auto-provisioned bricks, but it is also can be done with the same existing code. I still haven't convinced myself about the problem having both auto-provisioned and normal bricks in the same volume.

Having such flexibility really helps in features like snapshot, where snapshot bricks are auto-provisioned, but we have to support add/remove/replace/reset brick operations as normal. And this kind of operations becomes common when we have highly available snapshots. If bricks were missing at time taking snapshots, then we have to perform replace brick once we restored that snapshot.

So IMHO, strcture `type ProvisionState string`  should be a member of brick strcture.

**Comment by Madhu-1** [comment]
@rafikc30 , need to resolve conflicts.

 question:
in volume response is there any field for differentiating the volume is a normal or smart volume?

**Comment by rafikc30** [comment]
@Madhu-1 we have to add that field. I think we should add that field to brick response if we are having this struct in brickinfo.

**Comment by kshlm** [comment]
Agree with @phlogistonjohn about having a `ProvisionState` type. That gives us more ways to describe the provisioning method ("auto" or "snapshot clone" vs just "self-provisioned") and leaves room for expansion later on. Although I'm not a fan of adding fields to the base volinfo and brickinfo structs that aren't  absolutely necessary for normal functioning of a manually created gluster volume. I'd prefer that we do make use of `Volinfo.Metadata`, and add a `Brickinfo.Metadata` as well for such needs.

**Comment by rafikc30** [comment]
@aravindavk @kshlm @phlogistonjohn @Madhu-1 

I have refreshed the PR to address your comments.

**Comment by Madhu-1** [comment]
@rafikc30  there is an import cycle issue in code please fix it
```
import cycle not allowed in test
```

**Comment by aravindavk** [comment]
@rafikc30 do not add metadata through request struct. Set metadata in the Volume create handler. Volume create should not accept ineternal Metadata which starts with `_`(Currently it is accepted because validation is missing). Internal metadata keys are for plugins/components which can set Metadata key directly.

- Volume create request remains same as now
- Volume create handler sets internal Metadata based on `req.Size`

**Comment by rafikc30** [comment]
retest this please

File: glusterd2/bricksplanner/planner.go [hunk 1]
```diff
@@ -137,6 +141,9 @@ func getBricksLayout(req *api.VolCreateReq) ([]api.SubvolReq, error) {
 				TpMetadataSize: deviceutils.GetPoolMetadataSize(eachBrickTpSize),
 				FsType:         "xfs",
 				MntOpts:        "rw,inode64,noatime,nouuid",
+				Metadata: map[string]string{
```
Comment (aravindavk) [inline glusterd2/bricksplanner/planner.go]: This is still adding Metadata to request structure(All Metadata key starting with `_` need to be restricted via APIs). Add this directly to volinfo in the handler

File: pkg/utils/utils.go [hunk 2]
```diff
@@ -19,6 +19,45 @@ import (
...
+
+const (
+	//ProvisionKey is used to set the type of provisioning in volume/brick metadata
+	ProvisionKey string = "_brick-provision-type"
+
+	//ManuallyProvisioned bricks will be provisioned by a user
+	ManuallyProvisioned ProvisionState = ""
+	//AutoProvisioned bricks will be provsioned by device manager from glusterd2
+	AutoProvisioned ProvisionState = "auto"
+	//SnapshotProvisioned brciks will be created by performing a snapshot of a gluster brick
```
Comment (prashanthpai) [inline pkg/utils/utils.go]: typo

File: glusterd2/commands/volumes/volume-create-txn.go [hunk 3]
```diff
@@ -144,16 +147,23 @@ func newVolinfo(req *api.VolCreateReq) (*volume.Volinfo, error) {
...
-
 	if req.Metadata != nil {
 		volinfo.Metadata = req.Metadata
 	} else {
 		volinfo.Metadata = make(map[string]string)
 	}
 
+	if req.Size > 0 {
+		//AutoProvisioned volume
+		volinfo.Metadata[utils.ProvisionKey] = string(utils.AutoProvisioned)
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-create-txn.go]: Any reason why you chose `Metadata` over adding a field to `Volinfo` ?

File: pkg/utils/utils.go [hunk 4]
```diff
@@ -19,6 +19,45 @@ import (
 	log "github.com/sirupsen/logrus"
 )
 
+//ProvisionState is the way a brick or all bricks are provisioned
```
Comment (prashanthpai) [inline pkg/utils/utils.go]: Utils package is not a good place for this

Why not volume package ?

File: pkg/utils/utils.go [hunk 5]
```diff
@@ -19,6 +19,45 @@ import (
...
+	ManuallyProvisioned ProvisionState = ""
+	//AutoProvisioned bricks will be provsioned by device manager from glusterd2
+	AutoProvisioned ProvisionState = "auto"
+	//SnapshotProvisioned brciks will be created by performing a snapshot of a gluster brick
+	SnapshotProvisioned ProvisionState = "snapshot"
+)
+
+//IsManuallyProvisioned will return true if manually provisioned
+func (p ProvisionState) IsManuallyProvisioned() bool {
+	if p == ManuallyProvisioned {
```
Comment (prashanthpai) [inline pkg/utils/utils.go]: You can just `return p == ManuallyProvisioned`. Here and below

File: pkg/utils/utils.go (hunk 4 above)
Comment (rafikc30) [reply pkg/utils/utils.go]: ProvisionState is also used in brickinfo, that is why I decided to define in utils

File: glusterd2/commands/volumes/volume-create-txn.go (hunk 3 above)
Comment (rafikc30) [reply glusterd2/commands/volumes/volume-create-txn.go]: That was based on previous review comments

File: pkg/utils/utils.go (hunk 4 above)
Comment (prashanthpai) [reply pkg/utils/utils.go]: In that case, there can be two separate types - one in brick package and one in volume package.

**Comment by prashanthpai** [comment]
@kshlm @aravindavk Metadata, when introduced was meant for users/clients (and plugins) to store arbitrary data (akin to labels in k8s) which aren't interpreted by core glusterd2 in any way. In this specific case, how a volume is provisioned is an actionable internal information used by features that are core i.e volume management and snapshot. IMO, this is better stored as a field in `Volinfo` and `Brickinfo`. It need not be exposed in API though.

**Comment by atinmu** [comment]
@aravindavk @kshlm @prashanthpai can we get to a consensus on the approach?

**Pull request merged**